File: tests/qt_types_standalone/rust/src/qsize.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QSize;

#[cxx::bridge]
mod qsize_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/include/qt_types.h");

        type QSize = cxx_qt_lib::QSize;
    }

    extern "Rust" {
        fn construct_qsize() -> QSize;
        fn read_qsize(p: &QSize) -> bool;
        fn clone_qsize(p: &QSize) -> QSize;
    }
}

fn construct_qsize() -> QSize {
    QSize::new(1, 4)
}

fn read_qsize(s: &QSize) -> bool {
    s.width() == 1 && s.height() == 4
}

fn clone_qsize(s: &QSize) -> QSize {
    s.clone()
}

File: crates/cxx-qt-lib/src/core/qsize.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qsize.h"

#include <cxx-qt-lib/assertion_utils.h>

#include <cstdint>

// QSize has two "int" members
// Rust represents these as 32-bit integers.
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qsize.h?h=v5.15.6-lts-lgpl#n104
//
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qsize.h?h=v6.2.4#n113
assert_alignment_and_size(QSize, {
  ::std::int32_t a0;
  ::std::int32_t a1;
});

static_assert(::std::is_trivially_copyable<QSize>::value,
              "QSize must be trivially copyable!");
